Assistant: limit # lines returned in getFileContents tool

simonpcouch opened the pull request:
Part 1/3 of #10129 but does not close.
- Addresses https://github.com/posit-dev/positron/issues/10173

Rather than discouraging the use of the getFileContents tool due to potential high token usage, this PR implements a `lines` argument that prevents models from mistakenly reading the entire contents of large files.

- Add optional `lines` parameter to read specific line ranges from files
- Enforce 500-line limit when no line range is specified
- Return line range metadata (startLine, endLine, totalLines) in results and don't return `size`
- Remove `high-token-usage` tag from FileContentsTool
- Update assistant prompt to no longer discourage file contents retrieval via tools

Notably, the model _can_ bypass this limit by setting the `lines` argument with e.g. `[1, 100000]`. This is on purpose—I figure that a model likely knows if it does need to read the entire contents of a file, as in the example below.

Here's what this looks like in practice:

<img width="383" height="700" alt="Screenshot 2025-10-24 at 8 50 09 AM" src="https://github.com/user-attachments/assets/3137ff7c-044f-44d0-a471-66fc4d30a231" />

As of now, this failure stops the tool calling loop entirely—this will feel better in combination with https://github.com/posit-dev/positron/pull/10090, which will allow the model to "react to" the error.

Here's a full log of that interaction:  [read_large_file_failure.log](https://github.com/user-attachments/files/23124851/read_large_file_failure.log)


### Release Notes

<!--
  Optionally, replace `N/A` with text to be included in the next release notes.
  The `N/A` bullets are ignored. If you refer to one or more Positron issues,
  these issues are used to collect information about the feature or bugfix, such
  as the relevant language pack as determined by Github labels of type `lang: `.
  The note will automatically be tagged with the language.

  These notes are typically filled by the Positron team. If you are an external
  contributor, you may ignore this section.
-->

#### New Features

- N/A

#### Bug Fixes

- N/A


### QA Notes

<!--
  Positron team members: please add relevant e2e test tags, so the tests can be
  run when you open this pull request.

  - Instructions: https://github.com/posit-dev/positron/blob/main/test/e2e/README.md#pull-requests-and-test-tags
  - Available tags: https://github.com/posit-dev/positron/blob/main/test/e2e/infra/test-runner/test-tags.ts
-->


<!--
  Add additional information for QA on how to validate the change,
  paying special attention to the level of risk, adjacent areas that
  could be affected by the change, and any important contextual
  information not present in the linked issues.
-->

simonpcouch committed:
Assistant: limit # lines returned in getFileContents tool

- Add optional `lines` parameter to read specific line ranges from files
- Enforce 500-line limit when no line range is specified
- Return line range metadata (startLine, endLine, totalLines) in results and don't return `size`
- Remove `high-token-usage` tag from FileContentsTool
- Update assistant prompt to no longer discourage file contents retrieval via tools

samclark2015 reviewed (dismissed):
Looks good! I was able to confirm that by default, only a subset of files lines were read. I was able to trigger a full read with explicit instructions, too.

One oddity (perhaps unrelated to this PR): 
<img width="414" height="300" alt="Image" src="https://github.com/user-attachments/assets/3155677f-d93f-4a4c-8817-945f93ef0760" />

It successfully read the file in chunks prior to this message. But, after instructing to read it in one go, it hallucinated the file path (real path was `/Users/sclark/Projects/positron-playground/demo.ts`). This happened twice on two separate occasions during testing.

simonpcouch committed:
Merge branch 'main' into contents-10129

Signed-off-by: Simon P. Couch <simonpatrickcouch@gmail.com>

simonpcouch committed:
recreate prompt change with upstream

The same prompt now moved to a different file—make the same change to it as before.

simonpcouch commented:
Thanks for the review! Just merged upstream to make the same change to the prompt in the new file that it lives in.

I'm happy with this one being merged, but since I just merged upstream and yall have stale reviews turned on, I will defer to @samclark2015 or @sharon-wang to take one more look and merge if all is well. :)
